File: app/src/main/java/com/nikola/jakshic/dagger/data/remote/OpenDotaService.java
package com.nikola.jakshic.dagger.data.remote;

import com.nikola.jakshic.dagger.model.Hero;
import com.nikola.jakshic.dagger.model.Peer;
import com.nikola.jakshic.dagger.model.Player;
import com.nikola.jakshic.dagger.model.Competitive;
import com.nikola.jakshic.dagger.model.match.Match;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface OpenDotaService {

    String BASE_URL = "https://api.opendota.com/api/";

    @GET("search")
    Call<List<Player>> searchPlayers(@Query("q") String name);

    @GET("players/{account_id}/wl")
    Call<Player> getPlayerWinLoss(@Path("account_id") long playerId);

    @GET("proMatches")
    Call<List<Competitive>> getCompetitiveMatches();

    @GET("matches/{match_id}/")
    Call<Match> getMatch(@Path("match_id") long matchId);

    @GET("players/{account_id}/matches")
    Call<List<Match>> getMatches(
            @Path("account_id") long playerId,
            @Query("limit") long limit,
            @Query("offset") long offset);

    @GET("players/{account_id}/heroes")
    Call<List<Hero>> getHeroes(@Path("account_id") long playerId);

    @GET("players/{account_id}/peers")
    Call<List<Peer>> getPeers(@Path("account_id") long playerId);
}

File: app/src/main/java/com/nikola/jakshic/dagger/data/remote/OpenDotaService.kt
package com.nikola.jakshic.dagger.data.remote

import com.nikola.jakshic.dagger.vo.*
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface OpenDotaService {

    companion object {
        const val BASE_URL = "https://api.opendota.com/api/"
    }

    @GET("search")
    suspend fun searchPlayers(@Query("q") name: String): List<Player>

    @GET("players/{account_id}")
    suspend fun getPlayerProfile(@Path("account_id") id: Long): _Player

    @GET("players/{account_id}/wl")
    suspend fun getPlayerWinLoss(@Path("account_id") playerId: Long): Player

    @GET("players/{account_id}/matches?significant=0")
    suspend fun getMatches(
            @Path("account_id") playerId: Long,
            @Query("limit") limit: Int,
            @Query("offset") offset: Int): List<Match>

    @GET("players/{account_id}/matches?significant=0")
    suspend fun getMatchesByHero(
            @Path("account_id") playerId: Long,
            @Query("hero_id") heroId: Int,
            @Query("limit") limit: Int,
            @Query("offset") offset: Int): List<Match>

    @GET("players/{account_id}/heroes")
    suspend fun getHeroes(@Path("account_id") playerId: Long): List<Hero>

    @GET("players/{account_id}/peers")
    suspend fun getPeers(@Path("account_id") playerId: Long): List<Peer>

    @GET("matches/{match_id}/")
    suspend fun getMatch(@Path("match_id") matchId: Long): MatchStats

    @GET("proMatches")
    suspend fun getCompetitiveMatches(): List<Competitive>

    @GET("https://www.dota2.com/webapi/ILeaderboard/GetDivisionLeaderboard/v0001")
    suspend fun getLeaderboard(@Query("division") region: String): _Leaderboard
}